R/facet-labels.r

```diff
@@ -140,46 +267,109 @@ label_wrap_gen <- function(width = 25) {
 #' p2 %+% msleep +
 #'   facet_grid(vore ~ conservation2, labeller = labeller(vore = capitalize))
 #' p2 %+% msleep +
-#'  facet_grid(vore ~ conservation2, labeller = labeller(conservation2 =
-#'  label_wrap_gen(10)))
+#'   facet_grid(vore ~ conservation2,
+#'     labeller = labeller(conservation2 = label_wrap_gen(10))
+#'   )
 #' }
-labeller <- function(..., keep.as.numeric=FALSE) {
+#'
+#' # When multiple factors index the facetting, simply refer to one of
+#' # them to assign a labeller:
+#' p3 <- ggplot(mtcars, aes(wt, mpg)) + geom_point()
+#' p3 + facet_grid(gear + am ~ vs + cyl, labeller = labeller(
+#'   vs = label_bquote(.(y) ^ .(x)),
+#'   .default = label_value
+#' ))
+labeller <- function(..., keep.as.numeric = NULL, .as_character = FALSE,
+                     .multi_line = TRUE, .default = label_value) {
+  if (!is.null(keep.as.numeric)) {
+    .Deprecated(".as_character", old = "keep.as.numeric")
+  }
   args <- list(...)
 
-  function(variable, values) {
-    if (is.logical(values)) {
-      values <- as.integer(values) + 1
-    } else if (is.factor(values)) {
-      values <- as.character(values)
-    } else if (is.numeric(values) & !keep.as.numeric) {
-      values <- as.character(values)
-    }
+  function(labels) {
+    variable <- names(labels)
+    labels <- lapply(labels, function(values) {
+      if (is.logical(values)) {
+        as.integer(values) + 1
+      }
+      if (is.factor(values)) {
+        as.character(values)
+      } else if (is.numeric(values) && .as_character) {
+        as.character(values)
+      } else {
+        values
+      }
+    })
 
-    res <- args[[variable]]
+    labeller <- args[variable][[1]]
 
-    if (is.null(res)) {
-      # If the facetting margin (i.e. `variable`) was not specified when calling
-      # labeller, default to use the actual values.
-      result <- values
+    # If the facetting margin (i.e. `variable`) was not specified when
+    # calling labeller, default to use the actual values.
+    if (is.null(labeller)) {
```

Comment:
This is a legacy from before. We could remove the ability to use non-labeller functions that take and return a chr vector if you prefer?

Before the labeller function was checking the arguments to determine which functions are labellers, now I use the S3 class because I thought it was more robust. I can make it check the arguments as before if you think that'd be easier to create custom labellers. But we can't rely on `multi_line`, only on `labels`.